operator/v1/types_machineconfiguration.go

```diff
@@ -56,3 +63,78 @@ type MachineConfigurationList struct {
+	// +kubebuilder:validation:Required
+	// +required
+	MachineManagers []MachineManager `json:"machineManagers"`
+}
+
+// MachineManager contains identifying information of a machine management resource(eg. a machineset).
+// This is likely to evolve as support for more machine management resources are added.
+type MachineManager struct {
+	// resource is the machine management resource's type
+	// +kubebuilder:validation:Required
+	// +required
+	Resource MachineManagerResourceType `json:"resource"`
+	// apiGroup is name of the APIGroup that the machine management resource belongs to.
+	// +kubebuilder:validation:Required
+	// +required
+	APIGroup MachineManagerAPIGroupType `json:"apiGroup"`
+	// selection allows granular control of the machine management resources that will be enrolled to the ManagedBootImages feature.
+	// +kubebuilder:validation:Required
+	// +required
+	Selection MachineManagerSelector `json:"selection"`
+}
+
+// +kubebuilder:validation:XValidation:rule="has(self.mode) && self.mode == 'Partial' ?  has(self.partial) : !has(self.partial)",message="Partial is required when type is partial, and forbidden otherwise"
+// +union
+type MachineManagerSelector struct {
+	// mode is a union discriminator for MachineManagerSelector and can have three possible settings.
+	// When set to All, all machinesets are enrolled into the ManagedBootImages feature.
+	// When set to None, no machinesets are enrolled into the ManagedBootImages feature.
+	// When set to Partial, machinesets that match with label selector are enrolled into the ManagedBootImages feature.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	Mode MachineManagerSelectorMode `json:"mode"`
+
+	// partial provides a label selector that will be used by machine resources to enroll into the ManagedBootImages feature.
+	// Only permitted when mode is set to "Partial".
+	// +optional
+	Partial *metav1.LabelSelector `json:"partial,omitempty"`
+}
+
+// MachineManagerSelectorMode is a string enum used in the MachineManagerSelector union discriminator.
+type MachineManagerSelectorMode string
+
+const (
+	// All represents a configuration mode that will manage all machinesets.
+	All MachineManagerSelectorMode = "All"
+
+	// None represents a configuration that will not manage any machinesets.
+	None MachineManagerSelectorMode = "None"
+
+	// Partial represents a configuration mode that will manage machinesets specified by the label selector.
+	Partial MachineManagerSelectorMode = "Partial"
+)
+
+// MachineManagerSelectorMode is a string enum used in the MachineManagerSelector union discriminator.
+// +kubebuilder:validation:Enum:="machinesets"
+type MachineManagerResourceType string
+
+const (
+	// machinesets represent the MachineSet resource type, which manage a group of machines.
+	// Although this could belong to a MachineAPI or a ClusterAPI, only MAPI is currently supported.
+	MachineSets MachineManagerResourceType = "machinesets"
```

Comment:
If we were to add MachineDeployments, this is the enum for it, hence the somewhat abstract name. 
How about `MachineManagerManagedResourceType`?
```
const (
	// machinesets represent the MachineSet resource type, which manage a group of machines.
	// Although this could belong to a MachineAPI or a ClusterAPI, only MAPI is currently supported.
	MachineSets MachineManagerManagedResourceType = "machinesets"
        // machinedeployments represent the MachineDeployment resource type, which manage a group of machinesets, and is only prsent in ClusterAPI.
	MachineDeployments MachineManagerManagedResourceType = "machinedeployments"
)
```

Reply:
I appreciate that it's a little specific in my suggestion, however, this is just for the Go API, and the user facing API is not affected by the suggestion to have a more specific name in the Go types.

My concern is that we may want MachineSets or MachineDeployments in some other enum later, yet the consts should be distinguished as separate things

Reply:
Makes sense, I'll update to `MachineManagerMachineSetsResourceType`. Thanks!